pkg/manager/member/dm_worker_member_manager.go

```diff
@@ -0,0 +1,539 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/dmapi"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	"github.com/pingcap/tidb-operator/pkg/manager"
+	"github.com/pingcap/tidb-operator/pkg/util"
+
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	v1 "k8s.io/client-go/listers/apps/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	// dmWorkerDataVolumeMountPath is the mount path for dm-worker data volume
+	dmWorkerDataVolumeMountPath = "/var/lib/dm-worker"
+	// dmWorkerClusterCertPath is where the cert for inter-cluster communication stored (if any)
+	dmWorkerClusterCertPath = "/var/lib/dm-worker-tls"
+)
+
+type workerMemberManager struct {
+	masterControl dmapi.MasterControlInterface
+	setControl    controller.StatefulSetControlInterface
+	svcControl    controller.ServiceControlInterface
+	//podControl    controller.PodControlInterface
+	typedControl controller.TypedControlInterface
+	setLister    v1.StatefulSetLister
+	svcLister    corelisters.ServiceLister
+	podLister    corelisters.PodLister
+}
+
+// NewWorkerMemberManager returns a *ticdcMemberManager
+func NewWorkerMemberManager(masterControl dmapi.MasterControlInterface,
+	setControl controller.StatefulSetControlInterface,
+	svcControl controller.ServiceControlInterface,
+	//podControl controller.PodControlInterface,
+	typedControl controller.TypedControlInterface,
+	setLister v1.StatefulSetLister,
+	svcLister corelisters.ServiceLister,
+	podLister corelisters.PodLister) manager.DMManager {
+	return &workerMemberManager{
+		masterControl,
+		setControl,
+		svcControl,
+		//podControl,
+		typedControl,
+		setLister,
+		svcLister,
+		podLister,
+	}
+}
+
+func (wmm *workerMemberManager) Sync(dc *v1alpha1.DMCluster) error {
+	ns := dc.GetNamespace()
+	dcName := dc.GetName()
+
+	if dc.Spec.Worker == nil {
+		return nil
+	}
+	if dc.Spec.Paused {
+		klog.Infof("DMCluster %s/%s is paused, skip syncing dm-worker deployment", ns, dcName)
+		return nil
+	}
+	if !dc.MasterIsAvailable() {
+		return controller.RequeueErrorf("DMCluster: %s/%s, waiting for dm-master cluster running", ns, dcName)
+	}
```

Comment:
I think it's OK to remove this check if dm-worker can handle the case where dm-master service is unavailable.

Reply:
If dm-master doesn't have more than replicas/2 members alive, it can't offer service. I think we can keep it?